pysonde/thermodynamics.py

```diff
@@ -46,93 +48,141 @@ def convert_rh_to_dewpoint(temperature, rh):
+        raise ValueError(
+            f"Unknown method '{method}'. Available methods: 'hardy1998', 'wagner_pruss'."
+        )
+
+    # Convert result back to PintArray or xarray DataArray
+    if isinstance(temperature, pp.pint_array.PintArray):
+        pp.PintType.ureg = ureg
+        e_sw = pp.PintArray(e_sw, dtype="Pa")
+    elif isinstance(temperature, xr.DataArray) and hasattr(temperature.data, "_units"):
+        e_sw = xr.DataArray(
+            e_sw, dims=temperature.dims, coords=temperature.coords
+        ) * metpy.units.units("Pa")
+
+    return e_sw
 
 
 def calc_wv_mixing_ratio(sounding, vapor_pressure):
     """
-    Calculate water vapor mixing ratio
+    Works for both LEVEL-1 and LEVEL-2.
+
+    Key improvements:
+    - Handles both raw numeric values and Pint-quantified inputs.
+    - Dynamically detects pressure variable ("p" or "pressure").
+    - Ensures unit consistency across different dataset formats.
+    - Includes error handling for missing or incompatible units.
+
+    Parameters:
+    ----------
+    sounding : xarray.Dataset
+        Dataset containing pressure values.
+    vapor_pressure : xarray.DataArray or pint.Quantity
+        Vapor pressure values (must have units of Pascal).
+
+    Returns:
+    -------
+    xarray.DataArray
+        Water vapor mixing ratio with units of kg/kg.
     """
+
+    ureg = pint.get_application_registry()  # Ensure unit consistency
+
+    # Convert vapor_pressure to Pascal if necessary
     if isinstance(vapor_pressure, pp.pint_array.PintArray):
-        ureg = vapor_pressure.units._REGISTRY
-        vapor_pressure_Pa = vapor_pressure.quantity.to("Pa").magnitude
-    else:
-        vapor_pressure_Pa = vapor_pressure
-    if "pint" in sounding.pressure.dtype.__str__():
-        total_pressure = sounding.pressure.values.quantity.to("hPa").magnitude
+        vapor_pressure = vapor_pressure.quantity
+    elif isinstance(vapor_pressure, xr.DataArray):
+        try:
+            vapor_pressure = vapor_pressure.pint.to("Pa")
+        except AttributeError:
+            vapor_pressure = vapor_pressure.pint.quantify(
+                {"vapor_pressure": "Pa"}
+            ).pint.to("Pa")
+    elif isinstance(vapor_pressure, pint.Quantity):
+        vapor_pressure = vapor_pressure.to(ureg.Pa)
     else:
-        total_pressure = sounding.pressure.values
-    wv_mix_ratio = 1000.0 * (
-        (0.622 * vapor_pressure_Pa) / (100.0 * total_pressure - vapor_pressure_Pa)
+        vapor_pressure = vapor_pressure * ureg.Pa
+
+    # Identify and convert total pressure
+    pressure_var = (
+        "p" if "p" in sounding else "pressure" if "pressure" in sounding else None
     )
-    if isinstance(vapor_pressure, pp.pint_array.PintArray):
-        return wv_mix_ratio * ureg("g") / ureg("kg")
+    if pressure_var is None:
+        raise KeyError(
+            "No valid pressure variable found in the dataset! Expected 'p' or 'pressure'."
+        )
+
+    total_pressure = sounding[pressure_var]
```

Comment:
This should not be needed. pysonde should always use `pressure` internally. The mapping from e.g. `p` to `pressure` is done in the config files.